Segfault merging client configs

### Nomad version
Occurs in Nomad v0.5.4. Error below is from v0.5.2.

### Operating system and Environment details
Ubuntu 14.04

### Issue
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4bc7e6]

goroutine 1 [running]:
panic(0xfc0620, 0xc420012050)
	/opt/go/src/runtime/panic.go:500 +0x1a1
github.com/hashicorp/nomad/command/agent.(*ClientConfig).Merge(0xc4200f5a20, 0xc4200f5ad0, 0x1133dcc)
	/opt/gopath/src/github.com/hashicorp/nomad/command/agent/config.go:883 +0x816
github.com/hashicorp/nomad/command/agent.(*Config).Merge(0xc4200938c0, 0xc420093a20, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/command/agent/config.go:600 +0x147b
github.com/hashicorp/nomad/command/agent.LoadConfigDir(0x7ffc39f2cf5a, 0x11, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/command/agent/config.go:1166 +0x903
github.com/hashicorp/nomad/command/agent.LoadConfig(0x7ffc39f2cf5a, 0x11, 0x1, 0x1, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/command/agent/config.go:1085 +0x9b
github.com/hashicorp/nomad/command/agent.(*Command).readConfig(0xc42007f900, 0xc420056e70)
	/opt/gopath/src/github.com/hashicorp/nomad/command/agent/command.go:165 +0x12aa
github.com/hashicorp/nomad/command/agent.(*Command).Run(0xc42007f900, 0xc42000c110, 0x1, 0x1, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/command/agent/command.go:420 +0x16d
github.com/hashicorp/nomad/vendor/github.com/mitchellh/cli.(*CLI).Run(0xc4200ce0c0, 0xc4200ce0c0, 0x19, 0xc4200c3920)
	/opt/gopath/src/github.com/hashicorp/nomad/vendor/github.com/mitchellh/cli/cli.go:153 +0x330
main.RunCustom(0xc42000c100, 0x2, 0x2, 0xc4201164e0, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/main.go:55 +0x7a1
main.Run(0xc42000c100, 0x2, 0x2, 0xc4200001a0)
	/opt/gopath/src/github.com/hashicorp/nomad/main.go:20 +0x56
main.main()
	/opt/gopath/src/github.com/hashicorp/nomad/main.go:16 +0x66

### Reproduction steps
I had 1 config file with just enabled = true for the client block, with another config file that loaded after that with the reserved block inside the client block setting a reserved port range.  Adding an empty reserved block to the first config file fixed the issue.

--- a/command/agent/config.go
+++ b/command/agent/config.go
@@ -898,7 +898,10 @@ func (a *ClientConfig) Merge(b *ClientConfig) *ClientConfig {
 	if b.ClientMinPort != 0 {
 		result.ClientMinPort = b.ClientMinPort
 	}
-	if b.Reserved != nil {
+	if result.Reserved == nil && b.Reserved != nil {
+		reserved := *b.Reserved
+		result.Reserved = &reserved
+	} else if b.Reserved != nil {
 		result.Reserved = result.Reserved.Merge(b.Reserved)
 	}
 	if b.GCInterval != 0 {

--- a/command/agent/config_test.go
+++ b/command/agent/config_test.go
@@ -20,7 +20,21 @@ var (
 )
 
 func TestConfig_Merge(t *testing.T) {
+	c0 := &Config{}
+
 	c1 := &Config{
+		Telemetry:      &Telemetry{},
+		Client:         &ClientConfig{},
+		Server:         &ServerConfig{},
+		Ports:          &Ports{},
+		Addresses:      &Addresses{},
+		AdvertiseAddrs: &AdvertiseAddrs{},
+		Atlas:          &AtlasConfig{},
+		Vault:          &config.VaultConfig{},
+		Consul:         &config.ConsulConfig{},
+	}
+
+	c2 := &Config{
 		Region:                    "global",
 		Datacenter:                "dc1",
 		NodeName:                  "node1",
@@ -136,7 +150,7 @@ func TestConfig_Merge(t *testing.T) {
 		},
 	}
 
-	c2 := &Config{
+	c3 := &Config{
 		Region:                    "region2",
 		Datacenter:                "dc2",
 		NodeName:                  "node2",
@@ -271,9 +285,11 @@ func TestConfig_Merge(t *testing.T) {
 		},
 	}
 
-	result := c1.Merge(c2)
-	if !reflect.DeepEqual(result, c2) {
-		t.Fatalf("bad:\n%#v\n%#v", result, c2)
+	result := c0.Merge(c1)
+	result = result.Merge(c2)
+	result = result.Merge(c3)
+	if !reflect.DeepEqual(result, c3) {
+		t.Fatalf("bad:\n%#v\n%#v", result, c3)
 	}
 }
 

